llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -1585,6 +1585,26 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::MSTORE, VT, Custom);
     }
 
+    if (EnablePartialReduceNodes) {
+      for (MVT VT : MVT::integer_scalable_vector_valuetypes()) {
+        for (MVT InnerVT : MVT::integer_scalable_vector_valuetypes()) {
+          // 1. Set all combinations where a type is illegal to "Legal"
+          // - These will be legalized to a legal type pair
+          // - Avoid expanding them too early (or preventing folds)
+          if (!isTypeLegal(VT) || !isTypeLegal(InnerVT)) {
+            setPartialReduceMLAAction(VT, InnerVT, Legal);
+            continue;
+          }
+          //  2. Set all legal combinations to "Expand"
+          // - Not all of these can be lowered (via a Legal or Custom lowering).
+          setPartialReduceMLAAction(VT, InnerVT, Expand);
```

Comment:
There might be cases where during DAGCombine the type isn't legal, and after type legalisation the operation will still not be legal. In those cases, it might have been better to avoid using the PARTIAL_REDUCE_MLA nodes. What about marking everything as expand, except for ones where we know that legalisation will result in know will result in cases that are legal? (i.e. for types that are legal or require splitting, and where we know there is a DOT instruction for it)

Reply:
Yes, I did think about this issue when suggesting this (but left out trying to solve it for simplicity). I assumed that when producing partial reductions we'd only use types that are legal or we know will legalize to something that can lower to *DOT or wide adds.

Reply:
We might be able to use the information from this table to guide TargetTransformInfo on whether the operation is legal or not, and thus whether or not to create the node. So if this table is accurate, we can reuse that information in TTI.